lib/Sema/DerivedConformanceDifferentiable.cpp

```diff
@@ -56,9 +56,9 @@ getStoredPropertiesForDifferentiation(NominalTypeDecl *nominal,
   for (auto *vd : nominal->getStoredProperties()) {
     if (vd->getAttrs().hasAttribute<NoDerivativeAttr>())
       continue;
-    if (!vd->hasInterfaceType())
+    if (!vd->hasType())
```

Comment:
Why are you using the contextual type instead of the interface type everywhere?

Reply:
I copied this logic from [`DerivedConformanceEquatableHashable.cpp`](https://github.com/apple/swift/blob/8928431cbee1739e9834a5cf0dcee8809278358b/lib/Sema/DerivedConformanceEquatableHashable.cpp#L54) (the upstream code has changed slightly, but still uses contextual types).

I believe contextual types are necessary because they store conformance info that isn't available in interface types.

Example:
```
// vd->getValueInterfaceType(), no conformances
(generic_type_param_type depth=0 index=0 decl=simple.(file).GenericVectorSpacesEqualSelf.T@simple.swift:31:37)

// vd->getType(), has conformances
(archetype_type name=T address=0x7fdda8856500 conforms_to=Swift.(file).AdditiveArithmetic conforms_to=Swift.(file).Differentiable
  0x7fdda504ec00 Module name=simple
    0x7fdda8800c08 FileUnit file="simple.swift"
      0x7fdda880f168 StructDecl name=GenericVectorSpacesEqualSelf

  (nested_type=AllDifferentiableVariables <<unresolved>>)
  (nested_type=CotangentVector =T)
  (nested_type=TangentVector =T))
```

Reply:
On a related note, it definitely makes sense to move some duplicated helper `static` functions to a shared location (especially `allStoredPropertiesConformToProtocol`). [Here's a short list of functions that should be shared.](https://github.com/apple/swift/pull/21580#issuecomment-450713606)

@slavapestov: do you have any advice regarding this? My inclination is to move these functions to `CodeSynthesis.h`.